File: elixir/firsttry.erl
-module(firsttry).

-export([one/2, start/0, two/1]).

-spec firsttry:one(Other::pid(), Counter::number()) -> nil().
one(Other, 0) ->
    monitor(process, Other),
    Other ! {one, 0, self()},
    one(Other, 1);
one(Other, Cnt) ->
    timer:sleep(250),
    receive
      {Name, Counter} ->
        io:format("one received from ~w : ~w ~n", [Name, Counter]),
        Other ! {one, Cnt, self()},
        one(Other, Cnt + 1);
      Msg -> 
          io:format("one received ~w ~n", [Msg]),
          io:format("one exited~n")
    end.

%% two waits for a message from one, then respond. 10 times.
two(10) -> io:format("two exited~n");
two(Cnt) ->
    timer:sleep(250),
    receive
      {Name, Counter, From} ->
        io:format("two received from ~w : ~w ~n",
                [Name, Counter]),
        From ! {two, Counter}
    end,
    two(Cnt + 1).

start() ->
    spawn(fun () ->
        process_flag(trap_exit, true),
        Pid2 = spawn(firsttry, two, [0]),
        spawn_link(firsttry, one, [Pid2, 0]),
        receive
            Msg -> io:format("start ~w~n", [Msg])
        end,
        io:format("start done~n")
    end).

File: elixir/first-try.ex
defmodule FirstTry do

# one exchanges message with two and detect when two exits
  def one(other, 0) do
    Process.monitor(other)
    send(other, {:one, 0, self()})
    one(other, 1)
  end

  def one(other, cnt) do
    Process.sleep(250)
    receive do
      {name, counter} -> 
        IO.puts("one receive: #{name} @ #{counter}")
        send(other, {:one, cnt, self()})
        one(other, cnt+1)
      msg -> 
        IO.puts("one receive: #{inspect msg} EXIT!!!")
        IO.puts("one exited")
    end
  end

  def two(10) do
    IO.puts("two exited")
  end
  def two(cnt) do
    Process.sleep(250)
    receive do
      {name, counter, pid} -> 
        IO.puts("two receive: #{name} @ #{counter}")
        send pid, {"two", cnt}
    end
    two(cnt+1)
  end

end

pid2 = spawn(FirstTry, :two, [0])
spawn(FirstTry, :one, [pid2, 0])
